src/hotspot/share/runtime/objectMonitor.hpp

```diff
@@ -146,24 +165,12 @@ class ObjectMonitor : public CHeapObj<mtObjectMonitor> {
   // its cache line with _metadata.
   DEFINE_PAD_MINUS_SIZE(0, OM_CACHE_LINE_SIZE, sizeof(_metadata) +
                         sizeof(WeakHandle));
-  // Used by async deflation as a marker in the _owner field.
-  // Note that the choice of the two markers is peculiar:
-  // - They need to represent values that cannot be pointers. In particular,
-  //   we achieve this by using the lowest two bits.
-  // - ANONYMOUS_OWNER should be a small value, it is used in generated code
-  //   and small values encode much better.
-  // - We test for anonymous owner by testing for the lowest bit, therefore
-  //   DEFLATER_MARKER must *not* have that bit set.
-  static const uintptr_t DEFLATER_MARKER_VALUE = 2;
-  #define DEFLATER_MARKER reinterpret_cast<void*>(DEFLATER_MARKER_VALUE)
- public:
-  // NOTE: Typed as uintptr_t so that we can pick it up in SA, via vmStructs.
-  static const uintptr_t ANONYMOUS_OWNER = 1;
 
- private:
-  static void* anon_owner_ptr() { return reinterpret_cast<void*>(ANONYMOUS_OWNER); }
+  static const int64_t NO_OWNER = 0;
+  static const int64_t ANONYMOUS_OWNER = 1;
+  static const int64_t DEFLATER_MARKER = 2;
 
-  void* volatile _owner;            // pointer to owning thread OR BasicLock
+  int64_t volatile _owner;  // Either tid of owner, NO_OWNER, ANONYMOUS_OWNER or DEFLATER_MARKER.
   volatile uint64_t _previous_owner_tid;  // thread id of the previous owner of the monitor
```

Comment:
I was wondering what this was too but the _previous_owner_tid is the os thread id, not the Java thread id.

```
$ grep -r JFR_THREAD_ID
jfr/support/jfrThreadId.hpp:#define JFR_THREAD_ID(thread) (JfrThreadLocal::external_thread_id(thread))
jfr/support/jfrThreadId.hpp:#define JFR_THREAD_ID(thread) ((traceid)(thread)->osthread()->thread_id())
runtime/objectMonitor.cpp:    _previous_owner_tid = JFR_THREAD_ID(current);
runtime/objectMonitor.cpp:    iterator->_notifier_tid = JFR_THREAD_ID(current);
runtime/vmThread.cpp:  event->set_caller(JFR_THREAD_ID(op->calling_thread()));
```

Reply:
Then it looks like the JFR code needs updating as well, otherwise it is going to be reporting inconsistent information when virtual threads are locking monitors.

Reply:
So we use the os thread id when INCLUDE_JFR is not defined, but in that case we never actually post JFR events. So these _previous_owner_tid/_notifier_tid will be unused.